File: src/main/java/co/q64/stars/dimension/StarsDimension.java
package co.q64.stars.dimension;

import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.ChunkPos;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.World;
import net.minecraft.world.biome.Biome;
import net.minecraft.world.dimension.Dimension;
import net.minecraft.world.dimension.DimensionType;
import net.minecraft.world.gen.ChunkGenerator;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.api.distmarker.OnlyIn;

public abstract class StarsDimension extends Dimension {
    private ChunkGeneratorFactory generatorFactory;
    private World world;
    private Biome biome;
    private Vec3d color;

    protected StarsDimension(World world, DimensionType type, ChunkGeneratorFactory generatorFactory, Biome biome, Vec3d color) {
        super(world, type);
        this.world = world;
        this.biome = biome;
        this.generatorFactory = generatorFactory;
        this.color = color;
    }

    @Override
    protected void generateLightBrightnessTable() {
        for (int i = 0; i < lightBrightnessTable.length; i++) {
            lightBrightnessTable[i] = 1.0f;
        }
    }

    public ChunkGenerator<?> createChunkGenerator() {
        return generatorFactory.create(world);
    }

    public BlockPos findSpawn(ChunkPos chunkPosIn, boolean checkValid) {
        return chunkPosIn.asBlockPos();
    }

    public BlockPos findSpawn(int posX, int posZ, boolean checkValid) {
        return new BlockPos(posX, 42, posZ);
    }

    public float calculateCelestialAngle(long worldTime, float partialTicks) {
        return 0;
    }

    public boolean isSurfaceWorld() {
        return false;
    }

    public boolean isSkyColored() {
        return false;
    }

    @OnlyIn(Dist.CLIENT)
    @Override
    public Vec3d getSkyColor(BlockPos cameraPos, float partialTicks) {
        return color;
    }

    public Vec3d getFogColor(float celestialAngle, float partialTicks) {
        return color;
    }

    public boolean canRespawnHere() {
        return false;
    }

    public boolean doesXZShowFog(int x, int z) {
        return false;
    }

    public Biome getBiome(BlockPos pos) {
        return biome;
    }
}

File: src/main/kotlin/co/q64/stars/dimension/StarsDimension.kt
package co.q64.stars.dimension

import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.ChunkPos
import net.minecraft.util.math.Vec3d
import net.minecraft.world.IWorld
import net.minecraft.world.World
import net.minecraft.world.biome.Biome
import net.minecraft.world.dimension.Dimension
import net.minecraft.world.dimension.DimensionType
import net.minecraft.world.gen.ChunkGenerator
import net.minecraft.world.gen.GenerationSettings
import net.minecraftforge.api.distmarker.Dist
import net.minecraftforge.api.distmarker.OnlyIn


abstract class StarsDimension(
        world: World,
        type: DimensionType,
        private val generatorFactory: (IWorld) -> ChunkGenerator<GenerationSettings>,
        private val biome: Biome,
        private val color: Vec3d) : Dimension(world, type) {

    override fun generateLightBrightnessTable() {
        lightBrightnessTable.indices.forEach {
            lightBrightnessTable[it] = 1.0f
        }
    }

    override fun createChunkGenerator(): ChunkGenerator<*> = generatorFactory(world)
    override fun findSpawn(chunkPosIn: ChunkPos, checkValid: Boolean): BlockPos = chunkPosIn.asBlockPos()
    override fun findSpawn(posX: Int, posZ: Int, checkValid: Boolean) = BlockPos(posX, 42, posZ)
    override fun calculateCelestialAngle(worldTime: Long, partialTicks: Float) = 0.0f
    override fun isSurfaceWorld() = false
    override fun isSkyColored() = false
    override fun getFogColor(celestialAngle: Float, partialTicks: Float) = color
    override fun canRespawnHere() = false
    override fun doesXZShowFog(x: Int, z: Int) = false
    override fun getBiome(pos: BlockPos?) = biome

    @OnlyIn(Dist.CLIENT)
    override fun getSkyColor(cameraPos: BlockPos?, partialTicks: Float) = color
}